File: solutions/python/diffie-hellman/1/diffie_hellman.py
import secrets


def private_key(p):
    key = secrets.randbelow(p)
    return 2 if key < 2 else key


def public_key(p, g, private):
    return pow(g, private) % p


def secret(p, public, private):
    return pow(public, private) % p

File: csharp/diffie-hellman/DiffieHellman.cs
using System.Numerics;

public static class DiffieHellman
{
    public static BigInteger PrivateKey(BigInteger primeP) => 
        new(new Random().NextDouble() * (double) primeP);

    public static BigInteger PublicKey(BigInteger primeP, BigInteger primeG, BigInteger privateKey)
        => BigInteger.ModPow(primeG, privateKey, primeP);

    public static BigInteger Secret(BigInteger primeP, BigInteger publicKey, BigInteger privateKey) =>
        BigInteger.ModPow(publicKey, privateKey, primeP);
}